app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,284 @@
+        self.client = APIClient()
+        self.client.force_authenticate(self.user)
+
+    def test_retrieve_recipes(self):
+        """Test retrieving a list of recipes"""
+        sample_recipe(user=self.user)
+        sample_recipe(user=self.user)
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all().order_by('-id')
+        serializer = RecipeSerializer(recipes, many=True)
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_retrieve_recipes_limited_to_user(self):
+        """Test retrieve only the recipes of the given user"""
+        another_user = get_user_model().objects.create_user(
+            email='another@test.com',
+            password='testpass'
+        )
+        sample_recipe(user=another_user, title='another recipe')
+        sample_recipe(user=self.user)
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all().filter(user=self.user)
+        serializer = RecipeSerializer(recipes, many=True)
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(res.data), 1)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_view_recipe_detail(self):
+        """Test view recipe details"""
+        recipe = sample_recipe(user=self.user)
+        recipe.tags.add(sample_tag(user=self.user))
+        recipe.ingredients.add(sample_ingredient(user=self.user))
+
+        url = detail_url(recipe.id)
+        res = self.client.get(url)
+
+        serializer = RecipeDetailSerializer(recipe)
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_create_recipe(self):
+        """Test create a recipe"""
+        payload = {
+            'title': 'Bacon wrapped meatloaf',
+            'time_minutes': 60,
+            'price': 5.00
+        }
+        res = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        for key in payload.keys():
+            self.assertEqual(payload[key], getattr(recipe, key))
+
+    def test_create_recipe_with_tags(self):
+        """Test create a recipe with tags"""
+        tag1 = sample_tag(user=self.user, name='Meat')
+        tag2 = sample_tag(user=self.user, name='Main Course')
+        payload = {
+            'title': 'Bacon wrapped meatloaf',
+            "tags": [tag1.id, tag2.id],
+            'time_minutes': 60,
+            'price': 5.00
+        }
+
+        res = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        tags = recipe.tags.all()
+        self.assertEqual(tags.count(), 2)
+        self.assertIn(tag1, tags)
+        self.assertIn(tag2, tags)
+
+    def test_create_recipe_with_ingredients(self):
```

Comment:
Do we check that the user creating the recipe has access to the ingredients and tags linked in the POST payload by id?

Reply:
This is not done, the scenario described in the course is an optimistic case: we're creating only "good data" without thinking about what can go wrong. 
It should be something to care about in a real use case scenario though. I believe that a good name for such a technique is 'data segregation': basically it requires that you segregate data based on the possibility to access them.

Reply:
Yep exactly. If you are aware of the possible issues that's enough :)